pkg/resources/task.go

```diff
@@ -2,709 +2,648 @@ package resources
+			if err != nil {
+				return diag.FromErr(sdk.JoinErrors(err))
+			}
+			precedingTasks = append(precedingTasks, parentTaskId)
 		}
-		createRequest.WithAfter(precedingTasks)
+		req.WithAfter(precedingTasks)
 	}
 
 	if v, ok := d.GetOk("when"); ok {
-		createRequest.WithWhen(v.(string))
+		req.WithWhen(v.(string))
+	}
+
+	if parameterCreateDiags := handleTaskParametersCreate(d, req); len(parameterCreateDiags) > 0 {
+		return parameterCreateDiags
 	}
 
-	if err := client.Tasks.Create(ctx, createRequest); err != nil {
-		return fmt.Errorf("error creating task %s err = %w", taskId.FullyQualifiedName(), err)
+	if err := client.Tasks.Create(ctx, req); err != nil {
+		return diag.FromErr(err)
 	}
 
-	d.SetId(helpers.EncodeSnowflakeID(taskId))
+	// TODO(SNOW-1348116 - next pr): State upgrader for "id" (and potentially other fields)
+	d.SetId(helpers.EncodeResourceIdentifier(id))
 
-	enabled := d.Get("enabled").(bool)
-	if enabled {
-		if err := waitForTaskStart(ctx, client, taskId); err != nil {
-			log.Printf("[WARN] failed to resume task %s", name)
+	if d.Get("enabled").(bool) {
+		if err := waitForTaskStart(ctx, client, id); err != nil {
+			return diag.Diagnostics{
+				{
+					Severity: diag.Error,
+					Summary:  "Failed to start the task",
+					Detail:   fmt.Sprintf("Id: %s, err: %s", id.FullyQualifiedName(), err),
+				},
+			}
 		}
+		// Else case not handled, because tasks are created as suspended (https://docs.snowflake.com/en/sql-reference/sql/create-task; "important" section)
 	}
 
-	return ReadTask(d, meta)
-}
-
-func waitForTaskStart(ctx context.Context, client *sdk.Client, id sdk.SchemaObjectIdentifier) error {
-	err := client.Tasks.Alter(ctx, sdk.NewAlterTaskRequest(id).WithResume(true))
-	if err != nil {
-		return fmt.Errorf("error starting task %s err = %w", id.FullyQualifiedName(), err)
+	if err := client.Tasks.ResumeTasks(ctx, tasksToResume); err != nil {
+		log.Printf("[WARN] failed to resume tasks: %s", err)
 	}
-	return util.Retry(5, 5*time.Second, func() (error, bool) {
-		task, err := client.Tasks.ShowByID(ctx, id)
-		if err != nil {
-			return fmt.Errorf("error starting task %s err = %w", id.FullyQualifiedName(), err), false
-		}
-		if task.State != sdk.TaskStateStarted {
-			return nil, false
-		}
-		return nil, true
-	})
+
+	return ReadTask(false)(ctx, d, meta)
 }
 
-// UpdateTask implements schema.UpdateFunc.
-func UpdateTask(d *schema.ResourceData, meta interface{}) error {
+func UpdateTask(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
 	client := meta.(*provider.Context).Client
-	ctx := context.Background()
+	id, err := sdk.ParseSchemaObjectIdentifier(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
-	taskId := helpers.DecodeSnowflakeID(d.Id()).(sdk.SchemaObjectIdentifier)
+	task, err := client.Tasks.ShowByID(ctx, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
-	tasksToResume, err := client.Tasks.SuspendRootTasks(ctx, taskId, taskId)
-	defer func() {
```

Comment:
Why is this defer removed? In case of an error, we leave the previously suspended tasks as not resumed.

Reply:
Yeah, I left it for the next pr to handle all the cleanup cases correctly. For now, I tried multiple **multiple** times to make the resource run the actions in the correct order.